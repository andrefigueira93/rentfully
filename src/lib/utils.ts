import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

// Function to merge Tailwind classes
export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

// Function to format currency in a human readable format
export function formatCurrency(value?: number) {
  // If no value is provided, return NaN
  if (!value) {
    return "NaN";
  }

  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  }).format(value);
}

// Function to format date in a human readable format
export function humanDate(date: Date | undefined) {
  if (!date) {
    return "Invalid date";
  }

  return new Intl.DateTimeFormat("en-US", {
    dateStyle: "medium",
  }).format(date);
}

// Function to calculate the final price of a booking
export function calculateFinalPrice({
  price,
  extraPrice,
  guests = 1,
  totalDays = 1,
}: {
  price?: number;
  extraPrice?: number;
  guests?: number;
  totalDays?: number;
}) {
  // If any of the values are undefined, return NaN
  if (!price || !extraPrice) {
    return NaN;
  }

  const guestSum = guests > 1 ? guests - 1 : 1;
  const extraPriceSum = extraPrice * guestSum;
  const priceSum = price + extraPriceSum;
  const finalPrice = priceSum * totalDays;

  return finalPrice;
}
