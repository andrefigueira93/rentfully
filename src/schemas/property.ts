import z from "zod";

// Property schema
export const PropertySchema = z.object({
  id: z.number(),
  name: z.string().min(3).max(20),
  address: z.string().min(5),
  city: z.string().min(3),
  state: z.string().min(3),
  country: z.string().min(3),
  description: z.string().min(5),
  price: z.number().positive().min(5),
  extraGuestPrice: z.number().positive().min(5),
  maxGuests: z.number().positive().min(1),
  image: z
    .string()
    .optional()
    .default(`https://i.pravatar.cc/150?img=${Math.floor(Math.random() * 70)}`),
  visible: z.boolean(),
  timesRented: z.number().default(0),
  createdAt: z.date().default(() => new Date()),
  updatedAt: z.date().optional(),
});

// List Schema
export const ListPropertySchema = z.array(PropertySchema);

// While Creating, we don't need id, createdAt and updatedAt
export const CreatePropertySchema = PropertySchema.omit({
  id: true,
  timesRented: true,
  createdAt: true,
  updatedAt: true,
});

// While updating, we don't need createdAt and updatedAt
export const UpdatePropertySchema = PropertySchema.partial().omit({
  timesRented: true,
  createdAt: true,
  updatedAt: true,
});

// Exporting types for better type inference
export type PropertyProps = z.infer<typeof PropertySchema>;
export type ListPropertyProps = z.infer<typeof ListPropertySchema>;
export type CreatePropertyProps = z.infer<typeof CreatePropertySchema>;
export type UpdatePropertyProps = z.infer<typeof UpdatePropertySchema>;
