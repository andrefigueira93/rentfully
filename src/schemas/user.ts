import z from "zod";

// User schema
export const UserSchema = z.object({
  id: z.number(),
  firstName: z
    .string()
    .min(2, "Minimum of 2 characters")
    .max(20, "Maximum of 20 characters"),
  lastName: z
    .string()
    .min(2, "Minimum of 2 characters")
    .max(20, "Maximum of 20 characters"),
  email: z.string().email("Invalid email"),
  password: z.string().min(6).max(20),
  avatar: z
    .string()
    .optional()
    .default(`https://i.pravatar.cc/150?img=${Math.floor(Math.random() * 70)}`),
  confirmed: z.boolean(),
  createdAt: z.date().default(() => new Date()),
  updatedAt: z.date().optional(),
});

// List Schema
export const ListUserSchema = z.array(UserSchema);

// While Creating, we don't need id, createdAt and updatedAt
export const CreateUserSchema = UserSchema.omit({
  id: true,
  confirmed: true,
  createdAt: true,
  updatedAt: true,
});
// While updating, we don't need createdAt, updatedAt and confirmed
export const UpdateUserSchema = UserSchema.partial().omit({
  confirmed: true,
  createdAt: true,
  updatedAt: true,
});

// Exporting types for better type inference
export type UserProps = z.infer<typeof UserSchema>;
export type ListUserProps = z.infer<typeof ListUserSchema>;
export type CreateUserProps = z.infer<typeof CreateUserSchema>;
export type UpdateUserProps = z.infer<typeof UpdateUserSchema>;
