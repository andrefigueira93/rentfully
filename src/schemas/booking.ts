import z from "zod";
import { PropertyProps, PropertySchema } from "./property";

export const BookingSchema = z.object({
  id: z.number(),
  checkIn: z
    .date({
      invalid_type_error: "Invalid date format",
      required_error: "Check-in date is required",
      description: "Check-in date",
    })
    .min(new Date(), "Check-in date must be in the future"),
  checkOut: z.date({
    invalid_type_error: "Invalid date format",
    required_error: "Check-out date is required",
    description: "Check-out date",
  }),
  finalPrice: z.number().optional(),
  totalDays: z.number().optional(),
  propertyId: z.number().refine((val) => val, "Property is required"),
  userId: z.number().refine((val) => val, "User is required"),
  guests: z.number().int().min(1, "At least one guest is required"),
  status: z.enum(["Pending", "Confirmed", "Cancelled"]).default("Pending"),
  createdAt: z.date().default(() => new Date()),
  updatedAt: z.date().optional(),
});

export const CompleteBookingSchema = BookingSchema.extend({
  property: PropertySchema,
});

export const ListBookingSchema = z.array(BookingSchema);
export const CreateBookingSchema = BookingSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
});
export const UpdateBookingSchema = BookingSchema.partial().omit({
  id: true,
  createdAt: true,
  updatedAt: true,
});

// Exporting types for better type inference
export type BookingProps = z.infer<typeof BookingSchema>;
export type ListBookingProps = z.infer<typeof ListBookingSchema>;
export type CreateBookingProps = z.infer<typeof CreateBookingSchema>;
export type UpdateBookingProps = z.infer<typeof UpdateBookingSchema>;
export type CompleteBookingProps = z.infer<typeof CompleteBookingSchema>;
export type NewBookingProps = CreateBookingProps & {
  id: number;
  property?: PropertyProps;
};
