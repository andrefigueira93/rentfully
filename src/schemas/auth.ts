import z from "zod";
// Imported the UserSchema to reuse the already defined fields
import { UserSchema } from "./user";

//  Pick the email and password fields from the UserSchema
export const SignInSchema = UserSchema.pick({ password: true, email: true });

// Extend the SignInSchema with the terms field
export const SignUpSchema = UserSchema.pick({
  firstName: true,
  lastName: true,
  email: true,
  password: true,
}).extend({
  terms: z.boolean().refine((val) => val, {
    message: "You must accept the terms and conditions",
  }),
});

// Exporting types for better type inference
export type SignInProps = z.infer<typeof SignInSchema>;
export type SignUpProps = z.infer<typeof SignUpSchema>;
