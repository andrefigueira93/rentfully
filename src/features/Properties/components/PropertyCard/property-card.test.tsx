/* eslint-disable @typescript-eslint/no-explicit-any */
import { formatCurrency } from "@/lib/utils";
import { ListPropertyProps } from "@/schemas/property";
import { fireEvent, render, screen } from "@testing-library/react";
import { PropertyCard } from "./index";

const mockNewBooking = vi.fn();

vi.mock("@/hooks/useBooking", async () => {
  const actualModule = await vi.importActual("@/hooks/useBooking");
  return {
    ...(actualModule as any),
    useBooking: () => ({
      newBooking: mockNewBooking,
    }),
  };
});

describe("PropertyCard Component", () => {
  const mockProperty = {
    id: 1,
    name: "Test Property",
    address: "Test Address",
    city: "Test City",
    state: "Test State",
    country: "Test Country",
    image: "test-image.jpg",
    description: "Test Description",
    price: 100,
    visible: true,
    maxGuests: 5,
    extraGuestPrice: 50,
    timesRented: 5,
    createdAt: new Date(),
  } as ListPropertyProps[number];

  const mockScrollPosition = {
    x: 0,
    y: 0,
  };

  test("renders correctly with property data", () => {
    const { container } = render(
      <PropertyCard
        property={mockProperty}
        scrollPosition={mockScrollPosition}
      />
    );

    expect(screen.getByTestId("property-card")).toBeInTheDocument();
    expect(screen.getByText(mockProperty.name)).toBeInTheDocument();
    expect(screen.getByText(mockProperty.description)).toBeInTheDocument();
    expect(
      screen.getByText(`${mockProperty.timesRented} rents`)
    ).toBeInTheDocument();
    expect(
      screen.getByText(`${formatCurrency(mockProperty.price)}/night`)
    ).toBeInTheDocument();
    expect(screen.getByText("Book now")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("calls newBooking function on button click", () => {
    vi.mock("@/hooks/useBooking", () => ({
      useBooking: () => ({
        newBooking: mockNewBooking,
      }),
    }));

    const { container } = render(
      <PropertyCard
        property={mockProperty}
        scrollPosition={mockScrollPosition}
      />
    );

    const button = screen.getByText("Book now");
    fireEvent.click(button);

    expect(mockNewBooking).toHaveBeenCalledWith(mockProperty.id);
    expect(container).toMatchSnapshot();
  });
});
