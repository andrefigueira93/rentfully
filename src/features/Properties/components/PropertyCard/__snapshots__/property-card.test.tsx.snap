// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PropertyCard Component > calls newBooking function on button click 1`] = `
<div>
  <div
    class="variants defaultVariants border rounded-lg bg-white dark:bg-slate-900"
    data-testid="property-card"
  >
    <div
      class="flex-col space-y-1.5 p-6 flex md:flex-row space-x-6"
    >
      <span
        class=" lazy-load-image-background blur"
        style="color: transparent; display: inline-block; width: 100%;"
      >
        <img
          alt="Test Property"
          class="h-56 md:h-48 w-full rounded-md shadow"
          src="test-image.jpg"
          width="100%"
        />
      </span>
      <div
        class="flex flex-col w-full"
      >
        <h3
          class="font-semibold tracking-tight text-3xl"
        >
          Test Property
        </h3>
        <p>
          Test Address
          , 
          Test City
          , 
          Test Country
        </p>
        <p
          class="text-muted-foreground text-lg mt-4"
        >
          Test Description
        </p>
      </div>
      <div
        class="flex flex-col shrink-0"
      >
        <div
          class="inline-flex rounded-md border px-2.5 py-0.5 transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 justify-center self-end dark:text-white font-semibold text-lg items-center mb-4 md:mb-0"
        >
          <svg
            class="mr-1"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 2.75C8 2.47386 7.77614 2.25 7.5 2.25C7.22386 2.25 7 2.47386 7 2.75V7H2.75C2.47386 7 2.25 7.22386 2.25 7.5C2.25 7.77614 2.47386 8 2.75 8H7V12.25C7 12.5261 7.22386 12.75 7.5 12.75C7.77614 12.75 8 12.5261 8 12.25V8H12.25C12.5261 8 12.75 7.77614 12.75 7.5C12.75 7.22386 12.5261 7 12.25 7H8V2.75Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
           
          5
           rents
        </div>
        <div
          class="flex flex-col mt-auto justify-end"
        >
          <strong
            class="font-black text-xl mb-2"
          >
            $100.00
            /night
          </strong>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 font-bold dark:text-white"
          >
            Book now
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PropertyCard Component > renders correctly with property data 1`] = `
<div>
  <div
    class="variants defaultVariants border rounded-lg bg-white dark:bg-slate-900"
    data-testid="property-card"
  >
    <div
      class="flex-col space-y-1.5 p-6 flex md:flex-row space-x-6"
    >
      <span
        class=" lazy-load-image-background blur"
        style="color: transparent; display: inline-block; width: 100%;"
      >
        <img
          alt="Test Property"
          class="h-56 md:h-48 w-full rounded-md shadow"
          src="test-image.jpg"
          width="100%"
        />
      </span>
      <div
        class="flex flex-col w-full"
      >
        <h3
          class="font-semibold tracking-tight text-3xl"
        >
          Test Property
        </h3>
        <p>
          Test Address
          , 
          Test City
          , 
          Test Country
        </p>
        <p
          class="text-muted-foreground text-lg mt-4"
        >
          Test Description
        </p>
      </div>
      <div
        class="flex flex-col shrink-0"
      >
        <div
          class="inline-flex rounded-md border px-2.5 py-0.5 transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 justify-center self-end dark:text-white font-semibold text-lg items-center mb-4 md:mb-0"
        >
          <svg
            class="mr-1"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 2.75C8 2.47386 7.77614 2.25 7.5 2.25C7.22386 2.25 7 2.47386 7 2.75V7H2.75C2.47386 7 2.25 7.22386 2.25 7.5C2.25 7.77614 2.47386 8 2.75 8H7V12.25C7 12.5261 7.22386 12.75 7.5 12.75C7.77614 12.75 8 12.5261 8 12.25V8H12.25C12.5261 8 12.75 7.77614 12.75 7.5C12.75 7.22386 12.5261 7 12.25 7H8V2.75Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
           
          5
           rents
        </div>
        <div
          class="flex flex-col mt-auto justify-end"
        >
          <strong
            class="font-black text-xl mb-2"
          >
            $100.00
            /night
          </strong>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 font-bold dark:text-white"
          >
            Book now
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
